### ISSUE TITLE
Sinkhorn gets unstable results when executed on CPU.

### ISSUE THREAD

**Issue by zeakey** [t-vi/pytorch-tvmisc#9]
I got unstable results on CPU. The `get_coupling` function got inf values when executed on CPU.

**Comment by t-vi** [comment]
Thank you for the report. I'm afraid it's hard to follow up on this without a concrete example to reproduce this. Would you have one to share, please?

**Comment by zeakey** [comment]
@t-vi I created a gist to reproduce the error <https://gist.github.com/zeakey/c3ff4030f163b22a89e9ed287824678a>.

Actually I just simply move all computation from CUDA to CPU based on your original implementation.

**Comment by t-vi** [comment]
Indeed, there was a bug with the dimension for the logsumexp call
```
log_v[b] -= torch.logsumexp(-dist/lam+log_u[b, :, None], 0)
```
this previously had 1 as the second (= the dimension) argument but it needs to be 0 when there isn't a batch dimension after indexing that. Thank you for reporting this!

**Comment by zeakey** [comment]
@t-vi I got another problem with training a simple mnist classification model with Wasserstein loss.
The loss decreases as the training starts but the mode does not converge (accuracy stays around 10%).

Actually this is similar to the [Learning with a Wasserstein loss](http://cbcl.mit.edu/wasserstein/) paper and their code converges well on the MNIST dataset.

My code is here < https://gist.github.com/zeakey/c3ff4030f163b22a89e9ed287824678a>. Could you please have a look at it?

**Comment by zeakey** [comment]
Hi @t-vi , I tried to optimize two simple distribution with the wasserstein distance, and again I got very unstable results.

I simply optimize the wasserstein distance between two distributionsand update one of the distributions using gradient descent.

The full code is here, <https://gist.github.com/zeakey/c3ff4030f163b22a89e9ed287824678a>.
